pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,388 @@
+
+    >>> x = np.array(0.5, requires_grad=True)
+    >>> y = np.array([0.1, 0.2], requires_grad=True)
+    >>> print(qml.draw(circuit)(x, y))
+     0: ──H────────────────────╭Z──┤
+     1: ──RX(0.05)──RY(0.141)──╰C──┤ ⟨Z⟩
+
+    Evaluating the QNode, as well as the derivative, with respect to the gate
+    parameter *and* the transform weights:
+
+    >>> circuit(x, y)
+    0.9887793925354269
+    >>> qml.grad(circuit)(x, y)
+    (array(-0.02485651), array([-0.02474011, -0.09954244]))
+    """
+
+    if not isinstance(tape_transform, single_tape_transform):
+        raise ValueError("Can only convert single tape transforms into qfunc transforms!")
+
+    @functools.wraps(tape_transform)
+    def make_qfunc_transform(*targs, **tkwargs):
+        def wrapper(fn):
+            @functools.wraps(fn)
+            def internal_wrapper(*args, **kwargs):
+                tape = make_tape(fn)(*args, **kwargs)
+                tape = tape_transform(tape, *targs, **tkwargs)
+                return tape.measurements
+
+            return internal_wrapper
+
+        return wrapper
+
+    return make_qfunc_transform
+
+
+def qnode_transform(tape_transform):
+    """Register a new QNode transform.
+
+    Args:
+        tape_transform (tape transform): the tape transform
+            to convert into the QNode transform.
+
+            Allowed tape transforms must be functions of the following form:
+
+            .. code-block:: python
+
+                def tape_transform(tape, *args, **kwargs):
+                    ...
+                    return tapes, processing_fn
+
+            That is, the first argument must be the input tape to transform,
+            and the function must return a tuple ``(list, function)`` containing:
+
+            * A list of new tapes to execute, and
+            * A processing function with signature ``processing_fn(List[float])``
+              which is applied to the flat list of results from the executed tapes.
+
+    Returns:
+        function: The transformed QNode. Takes the same input arguments as
+        the input QNode.
+
+    **Example**
+
+    Given a simple tape transform, that replaces the :class:`~.CRX` gate with a
+    :class:`~.RY` and :class:`~.CZ` operation,
+
+    .. code-block:: python
+
+        @qml.single_tape_transform
+        def tape_transform(tape, x):
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    qml.RY(op.parameters[0] * qml.math.sqrt(x), wires=op.wires[1])
+                    qml.CZ(wires=op.wires)
+                else:
+                    op.queue()
+
+    lets build a QNode transform that applies this transform twice with different
+    transform parameters to create two tapes, and then summing the results:
+
+    .. code-block:: python
+
+        @qml.qnode_transform
+        def my_transform(tape, x, y):
+            tape1 = tape_transform(tape, x)

```

Comment:
It's actually so cool how you can have different parameters for the transforms. I imagine this could be used, e.g., to specify different bases to unroll in, or any special gate decompositions that would be used to override existing ones. :tada:

Reply:
Yeah! Also, I'm quietly excited for transforms that have continuous/differentiable parameters. Do any compilation routines come to mind? Most I know only take discrete parameters

Reply:
Don't know of any offhand, this could be interesting to investigate! :grin: